backfs.c

```diff
@@ -298,8 +298,15 @@ int backfs_read(const char *path, char *rbuf, size_t size, off_t offset,
 
         // in case another thread is reading a full block as a result of a 
         // cache miss
-        pthread_mutex_lock(&backfs.lock);
-        locked = true;
+        if (!locked)
+        {
```

Comment:
I just find that you didn't have this bug in the rw branch. However, your solution was to unlock at the end of every pass.  I will revert to this solution instead, it will ease the merging.